liquidhaskell-boot/src-ghc/Liquid/GHC/API/Extra.hs

```diff
@@ -170,46 +170,43 @@ apiCommentsParsedSource ps =
     spanToLineColumn =
       fmap (\s -> (srcSpanStartLine s, srcSpanStartCol s)) . srcSpanToRealSrcSpan
 
--- | Adds NOINLINE pragmas to all local bindings in the module.
+-- | Adds NOINLINE pragmas to all bindings in the module.
 --
 -- This prevents the simple optimizer from inlining such bindings which might
 -- have specs that would otherwise be left dangling.
 --
 -- https://gitlab.haskell.org/ghc/ghc/-/issues/24386
 --
-addNoInlinePragmasToLocalBinds :: ParsedModule -> ParsedModule
-addNoInlinePragmasToLocalBinds ps =
-    ps { pm_parsed_source = go (pm_parsed_source ps) }
+addNoInlinePragmasToBinds :: TcGblEnv -> TcGblEnv
+addNoInlinePragmasToBinds tcg = tcg{ tcg_binds = go False (tcg_binds tcg) }
   where
-    go :: forall a. Data a => a -> a
-    go = gmapT ((id `extT` addNoInlinePragmas) . go)
-
-    addNoInlinePragmas :: HsValBinds GhcPs -> HsValBinds GhcPs
-    addNoInlinePragmas = \case
-      ValBinds x binds sigs ->
-          ValBinds x binds (newSigs ++ dropInlinePragmas sigs)
-        where
-          dropInlinePragmas = filter (not . isInlineSig . unLoc)
-
-          isInlineSig InlineSig{} = True
-          isInlineSig _ = False
-
-          newSigs = concatMap (traverse noInlinePragmaForBind) $ bagToList binds
-
-          noInlinePragmaForBind :: HsBindLR GhcPs GhcPs -> [Sig GhcPs]
-          noInlinePragmaForBind bndr =
-              [ InlineSig
-                  []
-                  (noLocA b)
-                  defaultInlinePragma
-                    { inl_inline = NoInline $ SourceText $ occNameFS $ rdrNameOcc b
-                    , inl_act = NeverActive
-                    }
-              | b <- collectHsBindBinders CollNoDictBinders bndr
-              ]
-
-      bnds -> bnds
-
+    go :: forall a. Data a => Bool -> a -> a
+    go isAbs = gmapT $ go isAbs `extT` markHsBind isAbs
+
+    markHsBind :: Bool -> HsBind GhcTc -> HsBind GhcTc
+    markHsBind isAbs = \case
+        VarBind ext var rhs -> VarBind ext (markId isAbs var) (go False rhs)
+        FunBind ext var matches -> FunBind ext (markId isAbs <$> var) (go False matches)
+        PatBind ext lhs mult rhs -> PatBind ext lhs mult (go False rhs)
```

Comment:
Why are the ids bound by the lhs not marked?

Reply:
Good question. The GHC desugarer says:

```
        ; (force_var,sel_binds) <- mkSelectorBinds var_ticks pat PatBindRhs body'
          -- We silently ignore inline pragmas; no makeCorePair
          -- Not so cool, but really doesn't matter
```

So it was already a case before that any inline pragmas on these were ignored.

Reply:
That comment does not make explicit the purpose for which inline pragmas are ignored. And  the inline pragmas that `makeCorePair` reads are carried already in the Ids.

In addition, I have recollections of the NOINLINE pragmas making a difference to the simple optimizer.

Reply:
Of course the noinline pragmas, IN GENERAL, make a difference. THis whole patch is about replicating that!

But in the desugarer, `PatBind`s don't seem to take them into account.

I can try changing the logic so that we make all `PatBind`-bound `Id`s noinline as well, but I wouldn't be surprised if that starts breaking things instead of fixing anything.

Reply:
Here's an example where `NOINLINE` pragmas make a difference in pattern bindings.
```Haskell
module Map0 where                                                                                                                                                                             
                                                                                                                                                                                              
data Map k a = Tip                                                                                                                                                                            
             | Bin Int k a (Map k a) (Map k a)                                                                                                                                                
                                                                                                                                                                                              
deleteFindMax t
  = case t of
      Bin _ k x l Tip -> (k, x, l)
      Bin _ k x l r ->
        let {-# NOINLINE km3 #-}
            {-# NOINLINE vm #-}
            {-# NOINLINE rm #-}
            (km3, vm, rm) = deleteFindMax r
         in (km3, vm, (balance k x l rm))
      Tip           -> (error ms, error ms, Tip)
  where ms = "Map.deleteFindMax : can not return the maximal element of an empty Map"

balance :: k -> a -> Map k a -> Map k a -> Map k a                                                                                                                                            
balance k x l r = undefined
```
This is taken from the discussion in #2257.

I think we should mark ids in lhs to preserve such behavior.

Reply:
Please add this as a test today so I'll have it on `develop` tomorrow when I'll hope to finalize this PR

Reply:
Oh, my days are not as long as I want them, but I imagine such a test could go in
[GhcApiTests.hs](https://github.com/ucsd-progsys/liquidhaskell/blob/develop/liquidhaskell-boot/ghc-api-tests/GhcApiTests.hs).

The test could parse and typecheck the source code, then patch the bindings in the `TcgEnv`, then desugar, and then check that the bindings are indeed not inlined.